text/3301-atomic-memcpy.md

```diff
@@ -0,0 +1,227 @@
+- Feature Name: `atomic_memcpy`
+- Start Date: 2022-08-14
+- RFC PR: [rust-lang/rfcs#3301](https://github.com/rust-lang/rfcs/pull/3301)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+This is a proposal to add `AtomicPerByte<T>`, to represent _tearable atomics_.
+This makes it possible to properly implement a _sequence lock_ in Rust.
+
+# The Problem
+
+It's currently not possible to implement an efficient and perfectly
+(theoretically) correct sequence lock in Rust.
+
+Unlike most locking mechanisms, a sequence lock doesn't prevent a race
+to access the data it projects.
+Instead, it detects a race only after the load operation already happened,
+and retries it if the load operation raced with a write operation.
+
+A sequence lock in Rust looks something like this:
+
+```rust
+// Incomplete example
+
+pub struct SeqLock<T> {
+    seq: AtomicUsize,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl Sync<T: Copy + Send> for SeqLock<T> {}
+
+impl<T: Copy> SeqLock<T> {
+    /// Safety: Only call from one thread.
+    pub unsafe fn write(&self, value: T) {
+        self.seq.fetch_add(1, Relaxed);
+        write_data(&mut self.data, value, Release);
+        self.seq.fetch_add(1, Release);
+    }
+
+    pub fn read(&self) -> T {
+        loop {
+            let s1 = self.seq.load(Acquire);
+            let data = read_data(&self.data, Acquire);
+            let s2 = self.seq.load(Relaxed);
+            if s1 & 1 == 0 && s1 == s2 {
+                return unsafe { assume_valid(data) };
+            }
+        }
+    }
+}
+```
+
+The `write_data` and `read_data` calls can happen concurrently.
+The `write` method increments the counter before and after,
+such that the counter is odd during `write_data`.
+The `read` function will repeat `read_data` until
+the counter was identical and even both before and after reading.
+This way, `assume_valid` is only ever called on data that was
+not the result of a race.
+
+A big question is how to implement `write_data`, `read_data`, and `assume_valid`
+in Rust in an efficient way while satisfying the memory model.
+
+The somewhat popular `seqlock` crate and similar implementations found in the ecosystem
+all use a regular non-atomic write (preceded by an atomic fence) for writing,
+and `ptr::read_volatile` (followed by an atomic fence) for reading.
+This "works" "fine", but is technically undefined behavior.
```

Comment:
It may be worth noting that, besides from the volatile-volatile data race (which can be somewhat justified by always detecting and discarding the value of a racing read), the volatile-fence pattern doesn't actually provide synchronisations under the C++ memory model.

Fences only "interact" with atomic accesses. Volatiles are not atomic accesses. So unless SeqCst is involved, release fence-volatile write + volatile read-acquire fence does nothing synchronisation wise. This can cause the reader to miss a sequence update and use the volatile read value instead of discarding it, even though the writer did a racing write in the mean time.

Miri can already reproduce this: https://github.com/crossbeam-rs/crossbeam/issues/859. Though most (all?) LLVM compilation schemes emit the same code for volatile and relaxed: https://godbolt.org/z/MsGnq1fGh, so it probably doesn't blow up irl, until some optimisations come along